endpoint: move patching update functionality to pkg/endpoint

ianvernon opened the pull request:
A lot of Endpoint implementation information, including locking, internal fields, etc. was being leaked outside of the Endpoint package in the handling of `PATCH /endpoint`. Factor this into a function within the Endpoint package instead. As a result of this, the function `ForcePolicyCompute` is no longer used outside of `pkg/endpoint`, so do not export it.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8968)
<!-- Reviewable:end -->

Commit:
endpoint: factor out handling of patch operations to function in endpoint package

This removes the leakage of a lot of internal endpoint implementations
unnecessarily. Hide this in a function which the daemon can call instead.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: do not export ForcePolicyCompute function

This function is only used by `pkg/endpoint` now, do not export it.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani commented on daemon/endpoint.go:
```diff
@@ -385,14 +384,6 @@ func NewPatchEndpointIDHandler(d *Daemon) PatchEndpointIDHandler {
 	return &patchEndpointID{d: d}
 }
 
-func validPatchTransitionState(state models.EndpointState) bool {
```
I suppose the reason this was a standalone function was that PATCH was a specific operation outside of the endpoint. It might be more clear to rename this to something that reflects what we are really checking within the endpoint.

raybejjani replied:
Looking at the way we use this some more, I'm not sure what we gain by moving this function into pkg/endpoint. It really is specific to API itself and whether it allows something. If it were part of the endpoint then it could be part of ProcessChangeRequest (which might be better since the endpoint is then doing the check not the caller, and it already does it once).

ianvernon replied:
This is a symptom of the problem that we rely on endpoint's being on specific state in order to perform specific operations on them. I think this PR makes that problem more acute. I think the problem is that the daemon is aware of the endpoint's state at all. One of my goals here is to try to encapsulate the information about the endpoint away from callers, which includes its state. I am going to merge with that in mind, as this is more of an interim "let's make things better than they previously were" PR, which still has some weaknesses.

ianvernon replied:
Hm, I may be able to do this actually...

ianvernon replied:
Done!

raybejjani commented on pkg/endpoint/api.go:
```diff
@@ -0,0 +1,126 @@
+// purposes should a caller choose to try to regenerate this endpoint, as well
+// as an error if the Endpoint is being deleted, since there is no point in
+// changing an Endpoint if it is going to be deleted.
+func (e *Endpoint) ProcessChangeRequest(epTemplate *models.EndpointChangeRequest, newEp *Endpoint) (string, error) {
```
Should `endpoint.NewEndpointFromChangeModel` also move here since it's strictly an API related thing?

ianvernon replied:
Eventually, I'd like to do that. However, I'd have to then perform an endpointmanager lookup in the endpoint to see if the endpoint exists, which I can't do right now because the endpointmanager cannot be provided as a parameter anywhere.

raybejjani commented on pkg/endpoint/api.go:
```diff
@@ -0,0 +1,126 @@
+// purposes should a caller choose to try to regenerate this endpoint, as well
+// as an error if the Endpoint is being deleted, since there is no point in
+// changing an Endpoint if it is going to be deleted.
+func (e *Endpoint) ProcessChangeRequest(epTemplate *models.EndpointChangeRequest, newEp *Endpoint) (string, error) {
```
From what I can tell, we don't need epTemplate. All the fields are already filled in on newEP.

ianvernon replied:
@raybejjani can you PTAL again since I had to modify some of the actual functionality of the code in follow-up commits?

raybejjani commented on pkg/endpoint/api.go:
```diff
@@ -0,0 +1,126 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
```
A note in this file explaining that "api" in the filename is the cilium API and not the endpoint's API (or something) might be good.

raybejjani reviewed (approved):
Generally +1. I'm not very keen on ' ValidPatchTransitionState` but I'm not sure if you wanted to keep this a copy-paste refactor or were willing to change the order of operations.

Commit:
endpoint: do not pass endpoint model into `ProcessChangeRequest`

The `newEp` parameter has all the information we need; we don't need to pass in
the API model as well. As such, move the `ValidPatchTransitionState` function
out of `pkg/endpoint` back to `daemon` since the Endpoint package should not be
aware of what a `PATCH` operation is against it.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: clarify what `api.go` does

Signed-off by: Ian Vernon <ian@cilium.io>
